### PR TITLE
Add `Microsoft.Terminal.Remoting.dll` 

### PR DESCRIPTION
**Pull request by zadjii-msft**
Adds a `Microsoft.Terminal.Remoting.dll` to our solution. This DLL will
be responsible for all the Monarch/Peasant work that's been described in
#7240 & #8135. 

This PR does _not_ implement the Monarch/Peasant architecture in any
significant way. The goal of this PR is to just to establish the project
layout, and the most basic connections. This should make reviewing the
actual meat of the implementation (in a later PR) easier. It will also
give us the opportunity to include some of the basic weird things we're
doing (with `CoRegisterClass`) in the Terminal _now_, and get them
selfhosted, before building on them too much.

This PR does have windows registering the `Monarch` class with COM. When
windows are created, they'll as the Monarch if they should create a new
window or not. In this PR, the Monarch will always reply "yes, please
make a new window".

Similar to other projects in our solution, we're adding 3 projects here:
* `Microsoft.Terminal.Remoting.lib`: the actual implementation, as a
  static lib.
* `Microsoft.Terminal.Remoting.dll`: The implementation linked as a DLL,
  for use in `WindowsTerminal.exe`.
* `Remoting.UnitTests.dll`: A unit test dll that links with the static
  lib. 

There are plenty of TODOs scattered about the code. Clearly, most of
this isn't implemented yet, but I do have more WIP branches. I'm using
[`projects/5`](https://github.com/microsoft/terminal/projects/5) as my
notation for TODOs that are too small for an issue, but are part of the
whole Process Model 2.0 work.

## References

* #5000 - this is the process model megathread
* #7240 - The process model 2.0 spec.
* #8135 - the window management spec. (please review me, I have 0/3
  signoffs even after the discussion we had 😢)
* #8171 - the Monarch/peasant sample. (please review me, I have 1/2)

## PR Checklist
* [x] Closes nothing, this is just infrastructure
* [x] I work here
* [x] Tests added/passed
* [n/a] Requires documentation to be updated

### CODE REVIEW

File: build/Helix/runtests.cmd [hunk 1]
```diff
@@ -28,7 +28,7 @@ echo %TIME%
 powershell -ExecutionPolicy Bypass .\InstallTestAppDependencies.ps1
 echo %TIME%
 
-set testBinaryCandidates=TerminalApp.LocalTests.dll Conhost.UIA.Tests.dll
+set testBinaryCandidates=TerminalApp.LocalTests.dll SettingsModel.LocalTests.dll Conhost.UIA.Tests.dll
```
Comment (miniksa) [inline build/Helix/runtests.cmd:31]: I uh.... sneaking this in just to clean it up?

File: src/cascadia/Remoting/WindowManager.cpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    void WindowManager::_createMonarch()
+    {
+        // Heads up! This only works because we're using
+        // "metadata-based-marshalling" for our WinRT types. THat means the OS is
+        // using the .winmd file we generate to figure out the proxy/stub
+        // definitions for our types automatically. This only works in the following
+        // cases:
+        //
+        // * If we're running unpackaged: the .winmd but be a sibling of the .exe
```
Comment (miniksa) [inline src/cascadia/Remoting/WindowManager.cpp:77]: but be?

File: src/cascadia/Remoting/WindowManager.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+#include "Peasant.h"
+#include "Monarch.h"
+#include "../cascadia/inc/cppwinrt_utils.h"
+
+namespace winrt::Microsoft::Terminal::Remoting::implementation
+{
+    struct WindowManager : public WindowManagerT<WindowManager>
+    {
+        WindowManager();
+        ~WindowManager();
```
Comment (miniksa) [inline src/cascadia/Remoting/WindowManager.h:13]: destructor should probably be virtual or the class should probably be final.

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 4]
```diff
@@ -3,18 +3,18 @@
 
 #include "pch.h"
 #include "AppHost.h"
+// #include "MonarchFactory.h"
```
Comment (miniksa) [inline src/cascadia/WindowsTerminal/AppHost.cpp:6]: leftover?

File: tools/tests.xml [hunk 5]
```diff
@@ -5,6 +5,8 @@
   <test name="terminalCore" type="unit" binary="Terminal.Core.Unit.Tests.dll" />
   <test name="terminalApp" type="unit" binary="UnitTests_TerminalApp\Terminal.App.Unit.Tests.dll" />
   <test name="localTerminalApp" type="unit" runInHostApp="true" binary="TerminalApp.LocalTests.dll" />
+  <test name="localSettingsModel" type="unit" runInHostApp="true" binary="SettingsModel.LocalTests.dll" />
+  <test name="unitRemoting" type="unit" runInHostApp="true" binary="Remoting.UnitTests.dll" />
```
Comment (miniksa) [inline tools/tests.xml:9]: Wait, what? UnitTest runs in host app? Shouldn't this also be slated to run in Helix then?

**Review by miniksa — CHANGES_REQUESTED** [review]
Marking req changes so I can see what changed since I publish these. They're not super pressing comments, though.

File: build/Helix/runtests.cmd (hunk 1 above)
Comment (zadjii-msft) [reply build/Helix/runtests.cmd:31]: Yea, originally the remoting tests were local tests, and I noticed the SettingsModel local tests were missing in a bunch of places.

File: tools/tests.xml (hunk 5 above)
Comment (zadjii-msft) [reply tools/tests.xml:9]: derp, good catch

**Comment by zadjii-msft** [comment]
@DHowett you and I discussed being worried about peasants crashing after their original monarch is closed. Turns out, we don't need to worry about that in this PR! The monarch is only ever talked to once, at startup. After that, the window totally behaves as if business as usual.

**Review by miniksa — APPROVED** [review]

**Review by carlos-zamora — APPROVED** [review]
No issues with the implementation. Really you're just missing a few copyright headers haha

**Comment by DHowett** [comment]
Is the new project audit-enabled? Should it be?

File: src/cascadia/Remoting/MonarchFactory.h [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+#include "Monarch.h"
+
+// This seems like a hack, but it works.
+//
+// This class factory works so that there's only ever one instance of a Monarch
+// per-process. Once the first monarch is created, we'll stash it in g_weak.
+// Future callers who try to instantiate a Monarch will get the one that's
+// already been made.
+
+winrt::weak_ref<winrt::Microsoft::Terminal::Remoting::implementation::Monarch> g_weak{ nullptr };
```
Comment (DHowett) [inline src/cascadia/Remoting/MonarchFactory.h:15]: var storage locations in a header == bad news. importing the header multiple times (even accidentally) will cause linking failures

Comment (DHowett) [reply src/cascadia/Remoting/MonarchFactory.h:15]: make it a static inside the CreateInstance method perhaps

**Review by DHowett — COMMENTED** [review]
Moderately worried about the header things, but you can merge it if you plan to address it better later.

File: src/cascadia/Remoting/MonarchFactory.h [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            // Create a new Monarch instance
+            auto strong = winrt::make_self<winrt::Microsoft::Terminal::Remoting::implementation::Monarch>();
+
+            g_weak = (*strong).get_weak();
+            return strong.as(iid, result);
+        }
+        else
+        {
+            // We already instantiated one Monarch, let's just return that one!
```
Comment (DHowett) [inline src/cascadia/Remoting/MonarchFactory.h:39]: there is a tiny, tiny chance that g_weak goes from valid to invalid between line 29 and 40.

Robustness suggestion: we could try to lock g_weak and then check whether we _got_ an object.

Comment (zadjii-msft) [reply src/cascadia/Remoting/MonarchFactory.h:39]: okay that's a good suggestion

**Comment by zadjii-msft** [comment]
@msftbot don't merge this PR until 9pm

**Comment by zadjii-msft** [comment]
@msftbot wait 4 hours to merge this PR

File: src/cascadia/Remoting/Monarch.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+// Licensed under the MIT license.
+
+#pragma once
+
+#include "Monarch.g.h"
+#include "Peasant.h"
+#include "../cascadia/inc/cppwinrt_utils.h"
+
+// {06171993-7eb1-4f3e-85f5-8bdd7386cce3}
+constexpr GUID Monarch_clsid{
```
Comment (DHowett) [inline src/cascadia/Remoting/Monarch.h:11]:
```
56:#if defined(WT_BRANDING_RELEASE)
60:#elif defined(WT_BRANDING_PREVIEW)
```

**Comment by zadjii-msft** [comment]
@msftbot forget everything I just told you

**Pull request merged**